core/middlewares/with-routes.ts

```diff
@@ -252,8 +252,58 @@ const getRouteInfo = async (request: NextRequest, event: NextFetchEvent) => {
   }
 };
 
-export const withRoutes: MiddlewareFactory = () => {
+// See https://github.com/vercel/next.js/blob/main/packages/next/src/server/lib/server-action-request-meta.ts
+const isServerAction = ({ method, headers }: NextRequest) =>
+  method === 'POST' &&
+  (headers.get('next-action') != null ||
+    headers.get('content-type') === 'application/x-www-form-urlencoded' ||
+    headers.get('content-type')?.startsWith('multipart/form-data'));
+
+const getRewriteUrl = (
+  locale: string,
+  node: z.infer<typeof NodeSchema> | null | undefined,
+  request: NextRequest,
+  postfix: string,
+) => {
```

Comment:
Factored out a standalone piece of logic from `withRoutes` below to keep the `withRoutes` complexity under the limit. No semantic changes. I kept `RawHtmlPage` handling in `withRoutes` as a special case that is separate from the rewrite logic.

<img width="1117" alt="diff" src="https://github.com/user-attachments/assets/c3ad34d2-16af-4188-ad81-62149da15316">